### PR TITLE
plugin/transfer: Zone transfer plugin

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. Why is this pull request needed and what does it do?

New plugin *transfer* implementing the following parts of #3121
* AXFR
* Simple IXFR with AXFR fallback
* Define allowed transfer to hosts

NOTIFY is not implemented in this PR.

### 2. Which issues (if any) are related?
#3121 #2585

### 3. Which documentation changes (if any) need to be made?
Included

### 4. Does this introduce a backward incompatible change or deprecation?

### CODE REVIEW

**Comment by chrisohaver** [comment]
Status:
 * Basic AXFR structure in place.
 * Simple IXFR handling (fallback AXFR)

Todo:
 * Notify (in a following PR)

**Comment by chrisohaver** [comment]
I copied chunks of the transfer code from the file implementation. I see that #3227 modifies/reworks some of that.  I'll pull those changes in here when it merges.

**Comment by chrisohaver** [comment]
79d7b10 implements `transfer.Transferer` for *k8s_external*.  

This one was a bit tricky, as I had to extend the `external.Externaler` interface.  The result is a bit twisty. IMO, the code would be less verbose if we just allow *k8s_external* to use the kubernetes api connection directly instead of using a list of functions in the interface. That would make `external.Externaler` only implementable by plugins that have a k8s api connection... but thats a reasonable assumption.

File: plugin/transfer/transfer.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	Zones       []string
+	to          []string
+	Transferers map[string]Transferer // a map of zone->plugins which implement Transferer
+}
+
+// Transferer may be implemented by plugins to enable zone transfers
+type Transferer interface {
+	// AllRecords must return the SOA and all other records in the zone, including NS records
+	// for the given zone.
+	AllRecords(zone string) (*dns.SOA, []dns.RR, error)
```
Comment (miekg) [inline plugin/transfer/transfer.go]: This is, I think, not the correct interface. It needs to be more like an iterator.

```
type Transferer interface {
   Walk(zone string, serial uint32) ([]dns.RR, error)
}
~~~
Or something like that, see latest tweaks in `plugin/file/xfr.go`
Two things that:
1. First RR returned must be a SOA
2. `io.EOF` signals the end of the list

File: plugin/transfer/transfer.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+	// SoaSerial must return the current serial for the given zone.
+	SoaSerial(zone string) (uint32, error)
```
Comment (miekg) [inline plugin/transfer/transfer.go]: Probably can be folded into `Walk` as I did above

File: plugin/transfer/transfer.go (hunk 1 above)
Comment (johnbelamaric) [reply plugin/transfer/transfer.go]: What about returning a channel?

File: plugin/transfer/setup.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+								found = true
+								break
+							}
+						}
+						if !found {
+							continue
+						}
+						if x.Transferers[z] != nil {
+							// Acknowledge only the first transferer for each zone.
+							// This doesn't play nice with fallthrough, where more than one plugin can
```
Comment (johnbelamaric) [inline plugin/transfer/setup.go]: Would be nice to see if a later version could do something for this case.

File: plugin/transfer/transfer.go (hunk 1 above)
Comment (chrisohaver) [reply plugin/transfer/transfer.go]: Ah, a channel, so we can start responding to the client in parallel with the plugin generating the records.  Yeah, makes sense.

File: plugin/transfer/setup.go (hunk 3 above)
Comment (chrisohaver) [reply plugin/transfer/setup.go]: Yes, I thought about this and decided it was complicated enough to leave out for now, to keep this PR small and easily reviewable.  I think it's solvable.  The transfer plugin should be able to determine the "order" of plugins and merge them down with the records of other plugins later in the chain, preferring records from earlier plugins when there are collisions.  Multi instance plugins would need be handled by each plugin's implementation of transfer.

**Comment by chrisohaver** [comment]
I've reworked the interface, and I have adjusted the _k8s_external_ and _kubernetes_ implementations.

I'm inclined to just implement the interface for kubernetes/k8s_external in this PR, other plugins to be migrated in separate PRs.

File: plugin/kubernetes/controller.go [hunk 4]
```diff
@@ -43,13 +43,20 @@ type dnsController interface {
...
 
 type dnsControl struct {
-	// Modified tracks timestamp of the most recent changes
+	// modified tracks timestamp of the most recent changes
 	// It needs to be first because it is guaranteed to be 8-byte
 	// aligned ( we use sync.LoadAtomic with this )
 	modified int64
+	// extModified tracks timestamp of the most recent changes to
+	// services with external facing IP addresses
+    extModified int64
```
Comment (johnbelamaric) [inline plugin/kubernetes/controller.go]: go fmt needed?

File: plugin/kubernetes/controller.go [hunk 5]
```diff
@@ -423,6 +430,9 @@ func (dns *dnsControl) detectChanges(oldObj, newObj interface{}) {
 	switch ob := obj.(type) {
 	case *object.Service:
 		dns.updateModifed()
+		if len(ob.ExternalIPs) > 0 {
```
Comment (johnbelamaric) [inline plugin/kubernetes/controller.go]: can we check if the actual external IPs have changed?

File: plugin/kubernetes/xfr.go [hunk 6]
```diff
@@ -31,7 +31,7 @@ func (k *Kubernetes) Transfer(ctx context.Context, state request.Request) (int,
 
 	// Get all services.
 	rrs := make(chan dns.RR)
-	go k.transfer(rrs, state.Zone)
+	go k.transferAndClose(rrs, state.Zone)
```
Comment (johnbelamaric) [inline plugin/kubernetes/xfr.go]: So, after this merges, can we deprecate the transfer options within k8s, and move them all to transfer? Can we use transfer more than once with different zones so we can control the "allowed" IPs on a per-zone basis?

File: plugin/kubernetes/controller.go (hunk 5 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go]: Yeah - we also dont check if other Service attributes have changed... e.g.
* externalName, which would change CNAME records
* defined ports, which would change SRV records

ClusterIP is immutable, so don't need to check that...

File: plugin/kubernetes/xfr.go (hunk 6 above)
Comment (chrisohaver) [reply plugin/kubernetes/xfr.go]: Yes to both

File: plugin/kubernetes/controller.go (hunk 5 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go]: Added checks in f430772.  I tested manually, with a local cluster.
We don't have good unit test coverage in the controller.

File: plugin/kubernetes/controller.go [hunk 7]
```diff
@@ -510,6 +519,74 @@ func endpointsEquivalent(a, b *object.Endpoints) bool {
...
+		return true
+	}
+	for i := range oldSvc.ExternalIPs {
+		if oldSvc.ExternalIPs[i] != newSvc.ExternalIPs[i] {
+			return true
+		}
+	}
+
+	// All Port fields are mutable
+	for i := range oldSvc.Ports {
```
Comment (johnbelamaric) [inline plugin/kubernetes/controller.go]: you forgot to check the len on the ports this time, so if newPorts > oldPorts this can return a false negative.

**Comment by miekg** [comment]
I would prefer the entire xfr impl of k8s_external to be in another PR (I get the usefulness of testing your transfer impl, but it muddles this pr)

File: plugin/transfer/transfer.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	to          []string
+	Transferers map[string]Transferer // a map of zone->plugins which implement Transferer
+}
+
+// Transferer may be implemented by plugins to enable zone transfers
+type Transferer interface {
+	// AllRecords writes all records in the zone to the channel.  The
+	// first and last records written to the channel must be the SOA for the zone.
+	// If serial is non zero, an incremental zone transfer should be assumed.
+	AllRecords(zone string, serial uint32, ch chan dns.RR)
```
Comment (miekg) [inline plugin/transfer/transfer.go]: I don't like this interface yet. I think returning a channel is more natural, also the channel should be recieve only, not the send/receive it is now.
I.e. 
~~~
Transfer(zone string, serial uint32) (ch <-chan []dns.RR)
~~~
This also means the plugin closes the channel after is sent the last rrrs, which is more natural, not sure how that would work now.
And allowing a slice is also way more efficient, because that's probably what most plugins are holding anyway. Also sending the SOA twice seems axfr implemention detail and not super relevant to the implementation. The transfer plugin can take care of that.

File: plugin/transfer/transfer.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+		}
+	}
+	return false
+}
+
+// Name implements the Handler interface.
+func (Transfer) Name() string { return pluginName }
+
+const (
+	pluginName = "transfer"
```
Comment (miekg) [inline plugin/transfer/transfer.go]: meh, please just use "transfer" the few times it's need in the code.

File: plugin/transfer/transfer.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+	// Authoritative must return a list of Zones that the plugin is authoritative for.
+	Authoritative() []string
+}
+
+// ServeDNS implements the plugin.Handler interface.
+func (t Transfer) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+
+	var serial uint32
+	if state.QType() == dns.TypeIXFR {
```
Comment (miekg) [inline plugin/transfer/transfer.go:81]: this needs the ixfr handling that was recently added in file's tranfer

File: plugin/transfer/transfer.go [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+		soa, ok := r.Ns[0].(*dns.SOA)
+		if !ok {
+			return dns.RcodeServerFailure, nil
+		}
+		serial = soa.Serial
+	} else if state.QType() != dns.TypeAXFR {
```
Comment (miekg) [inline plugin/transfer/transfer.go]: `switch` is probably naturally

File: plugin/kubernetes/controller.go (hunk 7 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go]: oops. good catch.

File: plugin/transfer/transfer.go (hunk 10 above)
Comment (chrisohaver) [reply plugin/transfer/transfer.go:81]: I am currently handling this in the plugin implementations, to allow them to implement non-axfr-fallback if they could.  But implementing non-axfr-fallback is probably too much to expect of a plugin, and most plugins probably woudn't bother (the full ixfr answer protocol being a tad fiddly).  So, yeah , makes more practical sense to move the simple ixfr handling to the transfer plugin, and later we can extend the transfer plugin, perhaps with an IncrTransferer interface down the road.

**Comment by chrisohaver** [comment]
@miekg, I think I got all of your feedback thus far rolled in...

* remove _kubernetes_ and _k8s_external_ implementations from this PR
* use RR slice in the interface channel. create and close the channel in sender.
* use literal instead of constant for plugin name
* move simple ixfr response logic from plugin implementations to transfer plugin
* use a `switch` instead of `if` in ServeDNS

Let me know if you'd like me to squash the commits.

**Comment by miekg** [comment]
move the k8s transfer impl to a new pr.

Allrecords is oddly named and doesnt allow for incremental transfers to be
added, unless you go with some super weird naming scheme. SOA should not be
handled separately. And I don't think authoritative is needed as a method, you can just call Transfer and see what you get.
Probably best to let Transfer also return an predefined error or DNS rcode, to signal this

**Comment by chrisohaver** [comment]
> move the k8s transfer impl to a new pr.

@miekg, It's already removed from this PR ...  Do you want me to open a new PR in parallel before this one is finalized?

**Comment by chrisohaver** [comment]
@miekg, 

renamed AllRecords() -> Transfer() (per your earlier suggestion)
removed SOA() and Authoritative()

**Comment by johnbelamaric** [comment]
There some k8s change in there still - looks like maybe a bug fix?

**Comment by chrisohaver** [comment]
> There some k8s change in there still - looks like maybe a bug fix?

Sorry - i must have added it back in recently by accident ... trying to juggle the interdependent branches is challenging...

i'll rebase again...

**Comment by chrisohaver** [comment]
> Sorry - i must have added it back in recently by accident ... trying to juggle the interdependent branches is challenging...

No, it's something i missed when I untangled this PR from the k8s change.  Sorry about that.

**Comment by chrisohaver** [comment]
Actually, you are right - it was a bug fix... that was already merged in #3240 ...
not sure why it's coming up in my PR, my master is up to date, and I have rebased my branch with master ...

Maybe my local master branch is tainted.  I slipped off the git tightrope again.

Edit: Oh! Its the opposite of that bug fix... i.e. it's reverting the bug fix.  Not sure how I managed that...

**Comment by chrisohaver** [comment]
I have reconciled that k8s change with master.
Also added unit tests.

**Comment by johnbelamaric** [comment]
git is sharp and sometimes perplexing...

File: plugin/transfer/README.md [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+~~~
+transfer [ZONE...] {
+  to HOST...
+}
+~~~
+
+* **ZONES** The zones *transfer* will answer zone requests for. If left blank,
+  the zones are inherited from the enclosing server block. To answer zone 
+  transfers for a given zone, there must be another plugin in the same server
+  server block that serves the same zone, and implements `transfer.Transferer`.
```
Comment (johnbelamaric) [inline plugin/transfer/README.md]: server is repeated here

File: plugin/transfer/transfer.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+// Transfer is a plugin that handles zone transfers.
+type Transfer struct {
+	xfrs []*xfr
+	Next plugin.Handler
+}
+
+type xfr struct {
+	Zones       []string
+	to          []string
+	Transferers []Transferer // the list of plugins that implement Transferer
```
Comment (johnbelamaric) [inline plugin/transfer/transfer.go]: Why is this here instead of in the `Transfer` struct? In the setup you just add them all to every one anyway.

Comment (chrisohaver) [reply plugin/transfer/transfer.go]: ah, yes - it was there because it used to be a map of zones to Transferers ... only the zones for that instance were in it.  It can be moved to the main struct now.

Comment (chrisohaver) [reply plugin/transfer/transfer.go]: The list of Transferers is not likely to be large, and the load of transfer requests not likely to be high.  So the map optimization was not really needed.  Code is simpler without it.

**Comment by chrisohaver** [comment]
@miekg, @johnbelamaric 

Would you guys like to see **new PRs** for *kubernetes* and *k8s_external* plugin implementations of `transfer.Transfer` now or wait until after this PR is merged (if it merges)?

File: plugin/transfer/README.md [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+  block that serves the same zone, and implements `transfer.Transferer`.
+
+* `to ` **HOST...** The hosts *transfer* will transfer to. Use `*` to permit
+  transfers to all hosts.
+
+# Examples
+
+Answer zone transfers for `example.com.` using the theoretical plugin *plugin1*.
+
+~~~
```
Comment (miekg) [inline plugin/transfer/README.md]: this can use `corefile` is the language, so it's tested `plugin2` must be then renamed to an existing plugin.

File: plugin/transfer/README.md [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+## Description
+
+This plugin answers zone transfers for authoritative plugins that implement
+`transfer.Transferer`.
+
+Transfer answers AXFR requests and IXFR requests with AXFR fallback if the
+zone has changed.
+
+Notifies are not supported.
```
Comment (miekg) [inline plugin/transfer/README.md]: meaning... a plugin still needs to send those by itself? How does it know where to send them ?

File: plugin/transfer/README.md [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+*transfer* is not currently implemented by any plugins.
```
Comment (miekg) [inline plugin/transfer/README.md]: can be removed

File: plugin/transfer/README.md [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    plugin2 b.example.com
+}
+
+~~~
+
+Building on the example above, you can use *transfer* more than once in a server block to specify
+separate options per zone.
+
+~~~
+. {
```
Comment (miekg) [inline plugin/transfer/README.md]: this is not recommended, split these out on zone boundaries in diff. server blocks, the example itself also doesn't add much to the previous examples.

File: plugin/transfer/setup.go [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+				x.Zones[i] = plugin.Host(x.Zones[i]).Normalize()
+			}
+		} else {
+			x.Zones = make([]string, len(c.ServerBlockKeys))
+			for i := 0; i < len(c.ServerBlockKeys); i++ {
+				x.Zones[i] = plugin.Host(c.ServerBlockKeys[i]).Normalize()
+			}
+		}
+
+	BLOCK:
```
Comment (miekg) [inline plugin/transfer/setup.go]: s/BLOCK/block/

File: plugin/transfer/transfer.go (hunk 10 above)
Comment (miekg) [reply plugin/transfer/transfer.go:81]: Yep, that's was my thinking why `AllRecords` would be a bad name for the functions in the interfaces. Now we can extend this and keep a similar naming scheme

File: plugin/transfer/transfer.go [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+	return plugin.NextOrFailure(t.Name(), t.Next, ctx, w, r)
+
+}
+
+// send sends the response to the client.  It aborts if the first record received on the
+// channel is not an SOA.  send automatically adds the closing SOA to the end of an AXFR response.
+func (x xfr) send(state request.Request, fromPlugin <-chan []dns.RR) {
+	toRemote := make(chan *dns.Envelope)
+
+	go func() {
```
Comment (miekg) [inline plugin/transfer/transfer.go]: why is this done in a go-routine?

File: plugin/transfer/transfer.go [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+		var serial uint32
+		var soa *dns.SOA
+		l, c := 0, 0
+		for rs := range fromPlugin {
+			for _, r := range rs {
+				if c == 0 {
+					var ok bool
+					soa, ok = r.(*dns.SOA)
+					if !ok {
+						log.Errorf("First record in transfer from zone %s was not an SOA", state.QName())
```
Comment (miekg) [inline plugin/transfer/transfer.go]: return the error?

File: plugin/transfer/transfer.go [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+			toRemote <- &dns.Envelope{RR: records}
+		}
+
+		close(toRemote)
+		log.Infof("Outgoing transfer of %d records of zone %s to %s with %d SOA serial", c, state.QName(), state.IP(), serial)
+	}()
+
+	tr := new(dns.Transfer)
+	tr.Out(state.W, state.Req, toRemote)
+	state.W.Hijack()
```
Comment (miekg) [inline plugin/transfer/transfer.go]: hijack was removed from secondary because it's not thread safe, you're reintroducing it here.

File: plugin/transfer/transfer.go [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+		return dns.RcodeSuccess, nil
+	}
+
+	return plugin.NextOrFailure(t.Name(), t.Next, ctx, w, r)
+
+}
+
+// send sends the response to the client.  It aborts if the first record received on the
+// channel is not an SOA.  send automatically adds the closing SOA to the end of an AXFR response.
+func (x xfr) send(state request.Request, fromPlugin <-chan []dns.RR) {
```
Comment (miekg) [inline plugin/transfer/transfer.go]: the code is file/xfr.go is remarkbly shorter than this; why not use that?

File: plugin/transfer/transfer.go [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+	toRemote := make(chan *dns.Envelope)
+
+	go func() {
+		var records []dns.RR
+		var serial uint32
+		var soa *dns.SOA
+		l, c := 0, 0
+		for rs := range fromPlugin {
+			for _, r := range rs {
+				if c == 0 {
```
Comment (miekg) [inline plugin/transfer/transfer.go]: why do you introduce counters, and discard the counters from the `range`??

**Review by miekg — CHANGES_REQUESTED** [review]
interface looks good, but I have a question about sending notifies

File: plugin/transfer/README.md (hunk 14 above)
Comment (chrisohaver) [reply plugin/transfer/README.md]: I'll remove the examples for now, and add them back once when there are some plugins that implement transfer exist.

I'm hoping that this version of the README, where there are no plugins that implement transfer, will be short lived.

File: plugin/transfer/README.md (hunk 15 above)
Comment (chrisohaver) [reply plugin/transfer/README.md]: It's not implemented yet.  I was hoping to implement notify in a separate PR, to keep the PR short.  Is that OK, or do you want notifies to be handled in this PR too?

File: plugin/transfer/transfer.go (hunk 22 above)
Comment (chrisohaver) [reply plugin/transfer/transfer.go]: OK, the latest commit copies the code in from *file*, adapted to work with the channel returned by the plugin.

One functional difference I see is that *file*'s xfr implementation uses record count to cut/send envelopes, whereas previously I was using record data length. Not sure which one is more correct.

Another is that the code from *file* always sends the closing SOA in a separate envelope containing just that one record. Previously the closing SOA was included as the last record in the final envelope.

Not sure if these are significant.

File: plugin/transfer/transfer.go (hunk 21 above)
Comment (chrisohaver) [reply plugin/transfer/transfer.go]: removed

File: plugin/transfer/transfer.go (hunk 20 above)
Comment (chrisohaver) [reply plugin/transfer/transfer.go]: returns SERVFAIL with error now

File: plugin/transfer/transfer.go (hunk 23 above)
Comment (chrisohaver) [reply plugin/transfer/transfer.go]: code replaced/refactored

File: plugin/transfer/transfer.go (hunk 19 above)
Comment (chrisohaver) [reply plugin/transfer/transfer.go]: code replaced/refactored.

File: plugin/transfer/README.md (hunk 15 above)
Comment (johnbelamaric) [reply plugin/transfer/README.md]: What are you thinking here? How will `transfer` know the zone has changed? Periodic polling?

We could bring back the `watch` infrastructure, which allowed plugins to push an event down a channel ... though it would likely need some refactoring to work in this internal manner (`transfer` subscribing to updates of zones from other plugins, rather than clients subscribing to updates of names).

File: plugin/transfer/README.md (hunk 17 above)
Comment (johnbelamaric) [reply plugin/transfer/README.md]: I cannot really recommend always splitting in different server blocks until we fix the problems discussed in #3039. For any plugin that uses significant resources, duplicating it by having many server blocks is bad.

File: plugin/transfer/README.md (hunk 15 above)
Comment (chrisohaver) [reply plugin/transfer/README.md]: I was thinking polling because it seems to me like a simpler/cleaner solution.

In a polling design, a plugin would implement a `Serial(zone) serial` interface, which would be trivial to implement - just returning the current serial for the zone. Transfer would periodically poll each plugin in Tranferers for each zone, then send the notifies to listed hosts if serials change (as compared to a state maintained by the transfer plugin).

In a channel push design, a plugin would need to add code to their existing areas that update their serial.  This would not be as trivial to implement from the plugin perspective (code spread out a bit more).  There may be a cleaner way to do a push design that I'm not seeing right now.

Channel push would get notifies sent more timely, but how timely are notifies expected to be? It could also be more spammy, if thats a concern.

Comment (johnbelamaric) [reply plugin/transfer/README.md]: Yeah, that's a reasonable argument.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/transf..." ]
>+~~~
>+. {
>
>I cannot really recommend always splitting in different server blocks until we fix the problems discussed in #3039. For any plugin that uses significant resources, duplicating it by having many server blocks is bad.

Hooking it in "." is wrong. That this is still done in coredns/deployment for 
k8s is dumb as well.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/transf..." ]
>> +# transfer
>+
>+## Name
>+
>+*transfer* - answer zone transfers requests for compatible authoritative
>+plugins.
>+
>+## Description
>+
>+This plugin answers zone transfers for authoritative plugins that implement
>+`transfer.Transferer`.
>+
>+Transfer answers AXFR requests and IXFR requests with AXFR fallback if the
>+zone has changed.
>+
>+Notifies are not supported.
>
>Yeah, that's a reasonable argument.

so say, the *auto* plugin loads 30k small zones, how do we make that work? That 
needs thought. Dealing with notifies needs to fleshed out in this PR and 
possible a small doc needs to be written on how this will work.

File: plugin/transfer/transfer.go (hunk 22 above)
Comment (chrisohaver) [reply plugin/transfer/transfer.go]: > Not sure if these are significant.

Are they? (i.e. are these bugs in the *file* plugin's implementation of zone transfers, that I have now replicated here)

**Comment by chrisohaver** [comment]
> Hooking it in "." is wrong. That this is still done in coredns/deployment for
k8s is dumb as well.

Please see this discussion in coredns/deployment#153. Thanks.

**Comment by chrisohaver** [comment]
> so say, the *auto* plugin loads 30k small zones,

OK, to better support cases like that we could use channels instead.

Fleshing it out a bit, one way to do this ...

1. Create a _Notifier_ interface*, that contains one func, `Notify() <-chan  ZoneSerial`
which returns a channel to which the implementing plugin will send a `struct{Zone; Serial}` to the channel when a zone's serial changes.
2. In the On Startup caddy event, _transfer_ gets a channel from each plugin that implements Notify, and spins up listeners.
3. In each plugin, whenever the serial changes, sends an update to the channel.
4. When _transfer_ receives a message, it sends a NOTIFY to each static IP in the `to` option for the zone.

*If preferable, we could add `Notify()` to the `Transferer` interface, and permit plugins to return a nil channel (if they opt not to actually implement notify function).

**Comment by chrisohaver** [comment]
Gentle ping.

**Comment by chrisohaver** [comment]
@miekg, @johnbelamaric, please review my proposal for how we could add Notify support [above](https://github.com/coredns/coredns/pull/3223#issuecomment-534099365).

**Comment by johnbelamaric** [comment]
looks reasonable to me

**Comment by chrisohaver** [comment]
Thanks @johnbelamaric.  @miekg, is the notify concept [above](https://github.com/coredns/coredns/pull/3223#issuecomment-534099365) fleshed out enough, or would you like to see something more/different before merging this PR.

**Comment by chrisohaver** [comment]
FWIW - I've got notify + k8s_external working in locally in a separate set of branches - tested with a NIOS dns grid member as the secondary.  I added `Notify` to the existing `Transferer` interface, which I think is cleaner than to create a new interface for it. One change from the proposed solution [above](https://github.com/coredns/coredns/pull/3223#issuecomment-534099365) is that the message only contains zone (serial not needed when sending a notify).

**Comment by chrisohaver** [comment]
OK to merge?

File: plugin/transfer/transfer.go [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+type Transferer interface {
+	// Transfer returns a channel to which it writes responses to the transfer request.
+	// If the plugin is not authoritative for the zone, it should immediately return the
+	// Transfer.ErrNotAuthoritative error.
+	//
+	// If serial is 0, handle as an AXFR request. Transfer should send all records
+	// in the zone to the channel. The SOA should be written to the channel first, followed
+	// by all other records, including all NS + glue records.
+	//
+	// If serial is not 0, handle as an IXFR request. If the serial is >
```
Comment (johnbelamaric) [inline plugin/transfer/transfer.go]: Was that supposed to be `<`?

Comment (chrisohaver) [reply plugin/transfer/transfer.go]: ~yep thanks!~  actually it's supposed to be >=, and the next comparison <.  I got them swapped, thanks for drawing my attention to it.

**Comment by chrisohaver** [comment]
@miekg, I think all comments have been addressed and this is ready to go.  I'd like to merge this shortly unless you have any objections.
